src/osd/scrubber/osd_scrub_sched.h

```diff
@@ -122,193 +122,483 @@ ScrubQueue interfaces (main functions):
+	       : "dp";
+  }
+
+  bool is_periodic() const { return urgency <= urgency_t::overdue; }
+  bool is_viable() const { return urgency > urgency_t::off; }
+  bool is_scrubbing() const { return scrubbing; }
+
+  /**
+   * For sched-targets, lower is better.
+   * The <=> operator is used for "regular" comparisons.
+   * It assumes that both end of the comparison are not 'ripe'.
+   * But when sorting the scheduling queue - either for selecting the
+   * next job to be selected or for listing - we must take into account
+   * the 'ripeness' of the targets - which means we have to consult the
+   * clock. Do that efficiently - we use the 'eph_ripe_for_sort' flag.
    *
-   * @return a Scrub::attempt_t detailing either a success, or the failure
-   * reason.
+   * Note: 'partial order' due to strange utime_t::operator<=>()
    */
-  virtual schedule_result_t initiate_a_scrub(
-    spg_t pgid,
-    bool allow_requested_repair_only) = 0;
+  std::partial_ordering operator<=>(const SchedTarget&) const;
 
-  virtual ~ScrubSchedListener() {}
-};
+  bool operator==(const SchedTarget& r) const { return (*this <=> r) == 0; }
 
-}  // namespace Scrub
+  friend std::partial_ordering clock_based_cmp(
+      const SchedTarget& l,
+      const SchedTarget& r);
 
-/**
- * the queue of PGs waiting to be scrubbed.
- * Main operations are scheduling/unscheduling a PG to be scrubbed at a certain
- * time.
- *
- * A "penalty" queue maintains those PGs that have failed to reserve the
- * resources of their replicas. The PGs in this list will be reinstated into the
- * scrub queue when all eligible PGs were already handled, or after a timeout
- * (or if their deadline has passed [[disabled at this time]]).
- */
-class ScrubQueue {
- public:
-  enum class must_scrub_t { not_mandatory, mandatory };
-
-  enum class qu_state_t {
-    not_registered,  // not a primary, thus not considered for scrubbing by this
-		     // OSD (also the temporary state when just created)
-    registered,	     // in either of the two queues ('to_scrub' or 'penalized')
-    unregistering    // in the process of being unregistered. Will be finalized
-		     // under lock
-  };
+  bool is_ripe(utime_t now_is) const
+  {
+    return urgency > urgency_t::off && !scrubbing && now_is >= not_before;
+  }
 
-  ScrubQueue(CephContext* cct, Scrub::ScrubSchedListener& osds);
-  virtual ~ScrubQueue() = default;
+  void update_ripe_for_sort(utime_t now_is)
+  {
+    eph_ripe_for_sort = is_ripe(now_is);
+  }
 
-  struct scrub_schedule_t {
-    utime_t scheduled_at{};
-    utime_t deadline{0, 0};
-  };
+  bool over_deadline(utime_t now_is) const
+  {
+    return urgency > urgency_t::off && now_is >= deadline;
+  }
 
-  struct sched_params_t {
-    utime_t proposed_time{};
-    double min_interval{0.0};
-    double max_interval{0.0};
-    must_scrub_t is_must{ScrubQueue::must_scrub_t::not_mandatory};
-  };
+  // status
+  void set_scrubbing()
+  {
+    scrubbing = true;
+    push_nb_out(5s);
```

Comment:
Why do we push not_before out 5s when scrubbing?  I'm also a bit confused as to why we leave the job in the queue when scrubbing?